src/graphic/Fast3D/gfx_pc.cpp

```
@@ -2946,34 +2939,46 @@ bool gfx_movemem_handler_otr(F3DGfx** cmd0) {
     return false;
 }
 
-int16_t gfx_create_shader(const std::string& path) {
-    std::shared_ptr<Ship::ResourceInitData> initData = std::make_shared<Ship::ResourceInitData>();
-    initData->Path = path;
-    initData->IsCustom = false;
-    initData->ByteOrder = Ship::Endianness::Native;
-    auto shader = Ship::Context::GetInstance()->GetResourceManager()->GetArchiveManager()->LoadFile(path);
-    if (shader == nullptr || !shader->IsLoaded) {
-        return -1;
+int16_t gfx_search_existing_shader(const char* vertex, const char* fragment) {
+    ShaderMod shader = { vertex, fragment };
+    for (int i = 0; i < shader_ids.size(); i++) {
+        if (memcmp(&shader_ids[i], &shader, sizeof(ShaderMod)) == 0) {
+            return i;
+        }
     }
-    shader_ids.push_back(std::string(shader->Buffer->data()));
-    return shader_ids.size() - 1;
+    return -1;
 }
 
 bool gfx_set_shader_custom(F3DGfx** cmd0) {
     F3DGfx* cmd = *cmd0;
-    char* file = (char*)cmd->words.w1;
+    ShaderMod shader = { NULL, NULL };
+    shader.vertex = (char*)cmd->words.w1;
+    (*cmd0)++;
+    shader.fragment = (char*)((*cmd0)->words.w1);
 
-    if (file == nullptr) {
-        g_rdp.current_shader = { 0, 0, false };
-        return false;
+    if (shader.vertex != NULL || shader.fragment != NULL) {
+        // Search for duplicate shaders
+        auto cache = gfx_search_existing_shader(shader.vertex, shader.fragment);
+        if (cache != -1) {
+            g_rdp.current_shader = cache;
+        } else {
+            shader_ids.push_back(shader);
+            g_rdp.current_shader = shader_ids.size() - 1;
+        }
+    } else {
+        g_rdp.current_shader = -1;
     }
 
-    const auto path = std::string(file);
-    const auto shaderId = gfx_create_shader(path);
-    g_rdp.current_shader = { true, shaderId, (uint8_t)C0(16, 1) };
     return false;
 }
 
+std::optional<ShaderMod> gfx_get_shader(int16_t id) {
```

I don't like this use of `std::optional`, since it hides error cases from us. If an id is >= the size, or a negative value other than -1, that should error. Either an assert or a warning and we fall back to the default.

Consider an implementation that uses `uint16_t` as the id type, 0 is the default, and shader_id is seeded with the default shader. Each implementation calls `gfx_get_shader` and just uses the return value, not needing to check if it has a value. The logic of defaulting is handled by the invariant that 0 means default. If there is an error, we can log it and default to `shader_ids[0]`. Thoughts?